data/service/api/v1/datasets_data_create.go

```diff
@@ -106,7 +105,9 @@ func DataSetsDataCreate(dataServiceContext dataService.Context) {
 	for _, datum := range datumArray {
 		datum.SetUserID(dataSet.UserID)
 		datum.SetDataSetID(dataSet.UploadID)
-		datum.SetProvenance(CollectProvenanceInfo(ctx, req, details))
+		if authDetails.HasToken() {
+			datum.SetProvenance(CollectProvenanceInfo(ctx, req, authDetails))
+		}
```

Comment:
Might I suggest pushing all logic of what has and what does not have provenance into the `CollectProvenanceInfo` function?

```suggestion
		if provenance := CollectProvenanceInfo(ctx, req, authDetails) {
			datum.SetProvenance(provenance)
		}
```

Reply:
Then move the check for `HasToken` into `CollectProvenanceInfo`.

Reply:
> Might I suggest pushing all logic of what has and what does not have provenance into the `CollectProvenanceInfo` function?

Why? I don't understand how this is an improvement. You've not explained any reason for the change.

Reply:
Sure, I suppose I should have explained in more detail. 

It appears, at least to me, that there is a prerequisite required *before* calling `CollectProvenanceInfo`, namely ensuring that `authDetails.HasToken` is true. So, the function should either document that prerequisite for future use (by someone other than yourself) or it should handle the prerequisite itself. The latter seems the easiest and fully encapsulates the concept of "get provenance info from auth details".

Reply:
Thanks for clarifying. In going over your suggestion, I realized as well something that you wrote in your explanation:

> that there is a prerequisite required before calling CollectProvenanceInfo,

That was not the case previously. It wasn't my intention to make it a prerequisite, but you're right that my first edition of this PR did have that effect. I don't think that's desirable. I think it's better to collect whatever provenance information we can, from whatever's available. That's why I have tests to handle the case where any of the various values in the auth details are empty. The goal is to collect whatever information is there, and not abort the first time something that was expected is missing (though we should log that). Thanks for pointing that out. I'm going to change up my approach as a result.

Restating: The objective is to be callable with or without a token, but to not be a noisy logger unless the token isn't present *when it's expected to be present*. More concretely, if the request is not authenticated with a service secret, then we expect there to be a token. If there's not, that's something we should log. We shouldn't just return nil however. Doing so would break a bunch of tests, and just doesn't do any good.